# Assignment: Data Wrangling
### `! git clone https://github.com/ds4e/wrangling`
### Do Q1 and Q2, and either Q3 or Q4, for a total of 3 questions.

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)


2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing?
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [4]:
1. # Load csv file data into a pandas dataframe
import pandas as pd
!wget -O airbnb_hw.csv "https://raw.githubusercontent.com/dahalnirusma/wrangling/refs/heads/main/assignment/data/airbnb_hw.csv"
df = pd.read_csv("airbnb_hw.csv")
df.head()

--2025-02-03 22:26:13--  https://raw.githubusercontent.com/dahalnirusma/wrangling/refs/heads/main/assignment/data/airbnb_hw.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3414307 (3.3M) [text/plain]
Saving to: ‘airbnb_hw.csv’

airbnb_hw.csv       100%[===================>]   3.26M  --.-KB/s    in 0.08s   

2025-02-03 22:26:14 (39.1 MB/s) - ‘airbnb_hw.csv’ saved [3414307/3414307]



Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN  
3                  1                  0   199                   NaN  
4                  1                 39   549                  96.0

In [10]:
1. # Convert all strings in Price to numeric values
df["Price"] = pd.to_numeric(df["Price"], errors="coerce")
# Track total number of missing values
missing_values = df["Price"].isna().sum()
print(f"Missing values: {missing_values}")

Missing values: 181


The price column has no symbols so the next changes that should be made are converting the datatype of the values to numeric, (this is necessary when using the values from the column to perform arithmetic operations). This also converts all missing values into NaN. The total missing values is 181.

In [11]:
2. #Load csv file into pandas dataframe
!wget -O mn_police_use_of_force.csv "https://raw.githubusercontent.com/dahalnirusma/wrangling/refs/heads/main/assignment/data/mn_police_use_of_force.csv"
police_df = pd.read_csv("mn_police_use_of_force.csv")
df.head()

--2025-02-03 22:47:23--  https://raw.githubusercontent.com/dahalnirusma/wrangling/refs/heads/main/assignment/data/mn_police_use_of_force.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1873502 (1.8M) [text/plain]
Saving to: ‘mn_police_use_of_force.csv’

mn_police_use_of_fo 100%[===================>]   1.79M  --.-KB/s    in 0.07s   

2025-02-03 22:47:24 (24.4 MB/s) - ‘mn_police_use_of_force.csv’ saved [1873502/1873502]



Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews  Price  Review Scores Rating  
0                  1                  0  145.0                   NaN  
1                  1                  1   37.0                   NaN  
2                  1                  1   28.0                   NaN  
3                  1                  0  199.0                   NaN  
4                  1                 39  549.0                  96.0

In [13]:
# Check unique values and missing data
print(police_df['subject_injury'].value_counts(dropna=False))
print(police_df["subject_injury"].dtype)

subject_injury
NaN    9848
Yes    1631
No     1446
Name: count, dtype: int64
object


In [16]:
# Get missing values
missing_police_values = police_df["subject_injury"].isna().sum()
total_values = len(police_df)
missing_values_proportions = missing_values / total_values
print(f"Missing values: {missing_police_values}")
print(f"Proportion missing: {missing_values_proportions:.2%}")

# Cross-Tabulate
cross_tab = pd.crosstab(police_df["subject_injury"], police_df["force_type"], margins=True)
print(cross_tab)

Missing values: 9848
Proportion missing: 1.40%
force_type      Baton  Bodily Force  Chemical Irritant  Firearm  \
subject_injury                                                    
No                  0          1093                131        2   
Yes                 2          1286                 41        0   
All                 2          2379                172        2   

force_type      Gun Point Display  Improvised Weapon  Less Lethal Projectile  \
subject_injury                                                                 
No                             33                 34                       1   
Yes                            44                 40                       2   
All                            77                 74                       3   

force_type      Police K9 Bite  Taser   All  
subject_injury                               
No                           2    150  1446  
Yes                         44    172  1631  
All                         46  

2. 76.2% of the data is missing which is a huge concern as that means a significant part of the data has inconsistencies. A majority of the data missing means any conclusions derived from the data is skewed and biased. The tabulation reveals that missing values are not random as higher-force cases have a higher report rate versus low-force cases that have lower concentrations of reported injuries.

In [21]:
3.
import numpy as np
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
pretrial_df = pd.read_csv(url,low_memory=False)
pretrial_df.head()

# Check how the data is categorized
# print(pretrial_df["WhetherDefendantWasReleasedPretrial"].value_counts(dropna=False))
# print(pretrial_df["WhetherDefendantWasReleasedPretrial"].dtype)

# 1 - released pretrial, 0 - not released pretrial, 9 - missing or unknowndata
# replace the unknown columns with NaN
pretrial_df["WhetherDefendantWasReleasedPretrial"] = pretrial_df["WhetherDefendantWasReleasedPretrial"].replace(9, np.nan)
print(pretrial_df["WhetherDefendantWasReleasedPretrial"].value_counts(dropna=False))

WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
NaN       31
Name: count, dtype: int64


In [24]:
import numpy as np
4. # Cleaning ImposedSentenceAllChargeInContactEvent
print(pretrial_df["ImposedSentenceAllChargeInContactEvent"].value_counts(dropna=False))
print(pretrial_df["ImposedSentenceAllChargeInContactEvent"].dtype)

# Empty strings in the value counts
pretrial_df["ImposedSentenceAllChargeInContactEvent"] = pretrial_df["ImposedSentenceAllChargeInContactEvent"].replace(" ", np.nan)
print(pretrial_df["ImposedSentenceAllChargeInContactEvent"].isna().sum())

ImposedSentenceAllChargeInContactEvent
                    9053
0                   4953
12                  1404
.985626283367556    1051
6                    809
                    ... 
49.9712525667351       1
57.0349075975359       1
79.9260780287474       1
42.1642710472279       1
1.6570841889117        1
Name: count, Length: 484, dtype: int64
object
9053


4. The data has a significant number of values that are represented by an empty string. However, the emopty string does not properly state to people what those values are supposed to be/accurately capture that those are missing values. So to clean up and properly capture the different types of values in the columns, the empty string was replaced with np.Nan.

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks.

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?
7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.
8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?
9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)

In [29]:
import pandas as pd
# Import the shark attack data into a dataframe
shark_attacks = '/content/GSAF5.xls'
shark_attack_df = pd.read_excel(shark_attacks)
shark_attack_df.head()

#Drop columns without data
shark_attack_df = shark_attack_df.dropna(axis=1, how='all')

# Clean the year variable


Date    Year        Type        Country              State  \
0  2025-01-23 00:00:00  2025.0  Unprovoked      Australia  Western Australia   
1  2025-01-22 00:00:00  2025.0  Unprovoked      Australia  Western Australia   
2  2025-01-11 00:00:00  2025.0    Provoked            USA             Hawaii   
3  2025-01-02 00:00:00  2025.0  Unprovoked  New Caledonia       Grande Terre   
4  2025-01-02 00:00:00  2025.0  Unprovoked      Australia    South Australia   

                                  Location      Activity  \
0                           Shipwreck Cove       Bathing   
1                           Cheynes beach        Surfing   
2            Off Haleiwa Boat Harbour Oahu        Diving   
3              Islet of Kendek near Koumac  Spearfishing   
4  Granites Beach near Westall Streaky Bay       Surfing   

                                     Name Sex Age  ...  \
0                                    Boy    M  12  ...   
1                             Dale Kittow   M  37  ...   
2  Male not stated was a dive tour worker   M  23  ...   
3                        Robert Cuewapuru   M  40  ...   
4                           Lance Appleby   M  28  ...   

                                 Species   \
0                              Not stated   
1                           Bronze whaler   
2                              Not stated   
3          Reportedly Tiger or Bull shark   
4  Great White Shart estimated 4.8m 16 ft   

                                              Source  pdf href formula href  \
0                                     Bob Myatt GSAF  NaN          NaN  NaN   
1                                     Bob Myatt GSAF  NaN          NaN  NaN   
2                  Kevin McMurray Trackingsharks.com  NaN          NaN  NaN   
3                      Johannes Marchand Todd Smith   NaN          NaN  NaN   
4  Glen Folkard: Simon De Marchi News.com.au: The...  NaN          NaN  NaN   

  Case Number Case Number.1 original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN            NaN         NaN         NaN  
1         NaN           NaN            NaN         NaN         NaN  
2         NaN           NaN            NaN         NaN         NaN  
3         NaN           NaN            NaN         NaN         NaN  
4         NaN           NaN            NaN         NaN         NaN  

[5 rows x 23 columns]

**Q3.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Q4.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?